backend/gsr_booking/api_wrapper.py

```diff
@@ -532,38 +311,175 @@ def get_affiliation(self, email):
         else:
             return "Other"
 
-    def cancel_room(self, user, booking_id):
-        """Cancels room"""
-        gsr_booking = get_object_or_404(GSRBooking, booking_id=booking_id)
-        if gsr_booking:
-            if user != gsr_booking.user and user != gsr_booking.reservation.creator:
+            
+class BookingHandler:
+    def __init__(self, WBW=None, LBW=None):
+        self.WBW = WBW or WhartonBookingWrapper()
+        self.LBW = LBW or LibCalBookingWrapper()
+
+
+    def format_members(self, members):
+        return [
+            (User(**{ k[6:]: v for k, v in member.items() if k.startswith('user__') }), # temp user object
+            member['left'])
+            for member in members
+        ]
+
+    def get_wharton_members(self, group, gsr_id):
+        now = timezone.localtime()
+        ninty_min = datetime.timedelta(minutes=90)
+        zero_min = datetime.timedelta(minutes=0)
+        ret = GroupMembership.objects.filter(group=group, is_wharton=True).values('user').annotate(left=ninty_min-Coalesce(Sum(
+            F('user__gsrbooking__end')-F('user__gsrbooking__start'),
+            filter=Q(user__gsrbooking__gsr__gid=gsr_id) & Q(user__gsrbooking__is_cancelled=False) & Q(user__gsrbooking__end__gte=now)
+        ), zero_min)).filter(Q(left__gt=zero_min)).values('user__id', 'user__username', 'left').order_by('?')[:3]
+        return self.format_members(ret)
+
+        # locally i get this:
+        # SELECT "auth_user"."username", (5400000000 - COALESCE(SUM(django_timestamp_diff("gsr_booking_gsrbooking"."end", "gsr_booking_gsrbooking"."start")) FILTER (WHERE ("gsr_booking_gsr"."gid" = 1 AND NOT "gsr_booking_gsrbooking"."is_cancelled" AND "gsr_booking_gsrbooking"."end" >= 2023-10-22 20:04:00.804135)), 0)) AS "left" FROM "gsr_booking_groupmembership" LEFT OUTER JOIN "auth_user" ON ("gsr_booking_groupmembership"."user_id" = "auth_user"."id") LEFT OUTER JOIN "gsr_booking_gsrbooking" ON ("auth_user"."id" = "gsr_booking_gsrbooking"."user_id") LEFT OUTER JOIN "gsr_booking_gsr" ON ("gsr_booking_gsrbooking"."gsr_id" = "gsr_booking_gsr"."id") WHERE ("gsr_booking_groupmembership"."group_id" = 3 AND "gsr_booking_groupmembership"."is_wharton") GROUP BY "gsr_booking_groupmembership"."user_id", "auth_user"."username" HAVING (5400000000 - COALESCE(SUM(django_timestamp_diff("gsr_booking_gsrbooking"."end", "gsr_booking_gsrbooking"."start")) FILTER (WHERE ("gsr_booking_gsr"."gid" = 1 AND NOT "gsr_booking_gsrbooking"."is_cancelled" AND "gsr_booking_gsrbooking"."end" >= 2023-10-22 20:04:00.804135)), 0)) > 0 ORDER BY RAND() ASC LIMIT 3
+
+    def get_libcal_members(self, group):
+        day_start = timezone.localtime().replace(hour=0, minute=0, second=0, microsecond=0)
```

Comment:
nit: is there a `.date()` provided in `timezone` so we can avoid the `replace`?

Reply:
tuneer says its the best way